pkg/operator/options/options.go

```diff
@@ -57,113 +59,98 @@ type OptionFields struct {
+		return fmt.Errorf("failed to validate cli flags / env vars, invalid log level %q", o.LogLevel)
+	}
+	gates, err := FeatureGatesFromStr(o.FeatureGates.inputStr)
+	if err != nil {
+		return fmt.Errorf("failed to parse feature gates, %w", err)
 	}
+	o.FeatureGates = gates
 
-	o.FeatureGates = MustParseFeatureGates(o.FeatureGates.inputStr)
-
-	// Check if shared fields have been set. If they haven't, they may be ovewritten by settings parsed from configmaps.
-	o.FlagSet.Visit(func(f *flag.Flag) {
-		switch f.Name {
-		case "batch-max-duration":
-			o.BatchMaxDurationSet = true
-		case "batch-idle-duration":
-			o.BatchIdleDurationSet = true
-		case "feature-gates":
-			o.FeatureGatesSet = true
-		}
+	o.setFlags = make(map[string]bool)
+	fs.VisitAll(func(f *flag.Flag) {
+		envName := strings.ReplaceAll(strings.ToUpper(f.Name), "-", "_")
+		_, ok := os.LookupEnv(envName)
+		o.setFlags[f.Name] = ok
 	})
-	if _, ok := os.LookupEnv("BATCH_MAX_DURATION"); ok {
-		o.BatchMaxDurationSet = true
-	}
-	if _, ok := os.LookupEnv("BATCH_IDLE_DURATION"); ok {
-		o.BatchIdleDurationSet = true
-	}
-	if _, ok := os.LookupEnv("FEATURE_GATES"); ok {
-		o.FeatureGatesSet = true
-	}
+	fs.Visit(func(f *flag.Flag) {
+		o.setFlags[f.Name] = true
+	})
+
+	return nil
+}
 
-	return o, nil
+func (o *Options) ToContext(ctx context.Context) context.Context {
+	return ToContext(ctx, o)
 }
 
-// MergeSettings applies settings specified in the v1alpha5 configmap to options. If the value was already specified by
-// a CLI argument or environment variable, that value will be used.
-func (o *Options) MergeSettings(s *settings.Settings) *Options {
+// MergeSettings merges the settings in the context with the options in the context. Values already set in options take
+// precedent. If a value is specified in neither settings nor options, the default value is used.
+// Note: default values are applied to settings so its safe to use settings for unset option fields without knowing if
+// the value was set by the configmap.
+func (o *Options) MergeSettings(ctx context.Context) {
+	s := settings.FromContext(ctx)
 	if s == nil {
-		return o
+		return
 	}
 
-	// Note: settings also has default values applied to it. If the option is specified by neither Settings nor Options,
-	// the default value is used from Settings.
-	mergeField(&o.BatchMaxDuration, s.BatchMaxDuration, o.BatchMaxDurationSet)
-	mergeField(&o.BatchIdleDuration, s.BatchIdleDuration, o.BatchIdleDurationSet)
-	mergeField(&o.FeatureGates.Drift, s.DriftEnabled, o.FeatureGatesSet)
-	return o
+	if !o.setFlags["batch-max-duration"] {
+		o.BatchMaxDuration = s.BatchMaxDuration
+	}
+	if !o.setFlags["batch-idle-duration"] {
+		o.BatchIdleDuration = s.BatchIdleDuration
+	}
+	if !o.setFlags["feature-gates"] {
+		o.FeatureGates.Drift = s.DriftEnabled
+	}
 }
 
-func MustParseFeatureGates(gateStr string) FeatureGates {
+func FeatureGatesFromStr(gateStr string) (FeatureGates, error) {
```

Comment:
```suggestion
func ParseFeatureGates(gateStr string) (FeatureGates, error) {
```

nit: just a naming suggestion